### PR TITLE
Redirect on Cancel

### PR DESCRIPTION
**Pull request by LivioCavallo**
It would be useful to redirect to a user specified page after 'Cancel' of a new article submission.
At the moment it is not possible, we are redirected to the home page.

### Summary of Changes
Modify administrative com_content.ini file to signal user of new behaviour.
Modify components/com_content/controllers/article.php to enable redirect on cancel.

### Testing Instructions
Create menuitem of type Submit new article.

### Expected result
You would like to specify a redirect page after article submission cancel as you do in case of save/submit success.

### Actual result
You cannot.

### Documentation Changes Required
User indicated Redirection page will apply in case of Cancel too.

### CODE REVIEW

File: components/com_content/controllers/article.php [hunk 1]
```diff
@@ -159,9 +159,31 @@ protected function allowEdit($data = array(), $key = 'id')
...
+		{
+			$lang = '';
+
+			if (JLanguageMultilang::isEnabled())
+			{
+				$item = $app->getMenu()->getItem($menuitem);
+				$lang =  !is_null($item) && $item->language != '*' ? '&lang=' . $item->language : '';
+			}
+
+			// Redirect to the user spicified return page.
```
Comment (wojsmol) [inline components/com_content/controllers/article.php]: `specified` not `spicified`

Comment (LivioCavallo) [reply components/com_content/controllers/article.php]: I miss something. What do you mean?
The user specified return page is spicified by Itemid, retrieved from redirect_menuitem param.

Comment (wojsmol) [reply components/com_content/controllers/article.php]: There is a spelling error.

Comment (LivioCavallo) [reply components/com_content/controllers/article.php]: Oh! Thanks

**Comment by brianteeman** [comment]
I am wondering why we even have a redirect at all

**Comment by LivioCavallo** [comment]
Well, at the moment the behaviour is to go to the home page after cancel (1).
I think that letting the user to choose where to go (2) would be a big improvement.
Better options to achieve that than a redirect?

[Another usefull option would be (3) remembering the page where you were before opening the article submission page, and returning there after submission, cancel, ...  Adding this option would require other work; maybe a future PR ]

**Comment by brianteeman** [comment]
I agree that redirecting to the home page seems wrong. Personally if i am on a page with a form and i pressed cancel i would expect to be back at the empty form - my 2c

**Comment by LivioCavallo** [comment]
That seems a reasonable option too.
It could be easy obtained if we could have two different redirection targets: one for successful article submission, one for cancel. In case of cancel user could simply set to redirect to the same page.

It's easy to modify code to have two distinct redirection targets.
Do you think it could be useful? I could submit such a PR.

**Comment by brianteeman** [comment]
Personally I wouldnt have any option (we have too many) and I would just change the default behaviour to return to the form instead of the homepage on cancel - my 2c

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 84231f95051e98d50bd33e60e03c153235f312a7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17495">issues.joomla.org/tracker/joomla-cms/17495</a>.</sub>

**Comment by ghost** [comment]
+1 for PR. @brianteeman my Opinion is we have to low Options but to much Options unordered (Example "Options" of Articles are one long single Column).

**Comment by LivioCavallo** [comment]
I integrated some changes to include the behaviour suggested by @brianteeman as default option.
The additional option to set the cancel redirect page is shown just when requested by user (using showon in xml), so mantaining interface as clean as possible.

_[testing this PR, I detected a redirection malfunction in this class, save function. I'll get a PR ready. Should I wait tu submit the PR till this is merged or closed, o I'd better submitting the new PR right now?]_

File: components/com_content/controllers/article.php [hunk 2]
```diff
@@ -160,9 +160,63 @@ public function cancel($key = 'a_id')
...
+			$lang = '';
+
+			if ($menuitemId > 0)
+			{
+				$lang = '';
+				$item = $app->getMenu()->getItem($menuitemId);
+
+				if (JLanguageMultilang::isEnabled())
+				{
+					$lang =  !is_null($item) && $item->language != '*' ? '&lang=' . $item->language : '';
```
Comment (Quy) [inline components/com_content/controllers/article.php]: Remove space after `=`

**Comment by zero-24** [comment]
@LivioCavallo with the last two commits drone should be happy now 😄 

> [testing this PR, I detected a redirection malfunction in this class, save function. I'll get a PR ready. Should I wait tu submit the PR till this is merged or closed, o I'd better submitting the new PR right now?]

As this is a different method and has nothing to do with this PR i would suggest to do a new PR just containing the change you suggest. Together with a description what you are trying to fix and how to reproduce that 😄

**Comment by LivioCavallo** [comment]
More infos: the proposed solution integrates a workaround to erroneous route building. The problem is explained here #17599 and here #16774.
In details, at rows 185 and 190 I use the item's link property in addition to Itemid to build redir link and then send it to JRoute, something like:  $redirlink = $item->link . $lang . '&Itemid=' . $cancelMenuitemId;  passed to JRoute.
If you use just the Itemid the problems described in the above mentioned Issues arise.

Are the above mentioned issues solved in J!3.8 ? Will this code be a clean solution also in J!3.8?

_[I won't make a PR to fix 'save' function as it now seems related to  #17599 and #16774; solving that will solve save function problems too]_

**Comment by LivioCavallo** [comment]
Please, check my English in language .ini file...

**Comment by Kev1n337** [comment]
I have tested this item :white_check_mark: successfully on 0f282d5c44ac131601137855113c0b235229ad72<br /><br />Test before applying patch:
* Added menu item "create article"
* No submission page: -> redirect to home page
* Specified submission redirected: -> redirected on save, but not on cancel

After applying patch:
* No submission redirect: -> redirect to home page
* Specified submission redirected: -> redirected on save and on cancel
* Also tested translation successfully<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17495">issues.joomla.org/tracker/joomla-cms/17495</a>.</sub>

@icampus

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 0f282d5c44ac131601137855113c0b235229ad72<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17495">issues.joomla.org/tracker/joomla-cms/17495</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Pull request merged**